### PR TITLE
[DS-3989] Always capture curation task output.

### PR DESCRIPTION
**Pull request by mwoodiupui**
https://jira.duraspace.org/browse/DS-3989
See also https://wiki.duraspace.org/display/~terrywbrady/Curation+System+Needs

Make reporting a bit more abstract, so that different triggers can handle reports differently.
Give more flexibility in reporting than "standard output or nothing".
Capture all reports, not just the last one.

Passes manual CLI testing and integration test.

### CODE REVIEW

**Comment by terrywbrady** [comment]
It is nice to see a start on this especially if that start could be fairly simple.

I see that the print writer is used.  Are there any limitations to the output format that could be produced (TXT, JSON, other)

Given the implementation, how could the following be accomplished
- Would the UI be able to instruct the curation task that output should be preserved?  Could the UI then retrieve that ouput?
- Could a CLI invocation instruct the curation task to preserve output?  Will the CLI task know the resulting output file path?
- If either CLI or UI wants to defer a curation task to the queue, how would that change the serialized ouput?

**Comment by mwoodiupui** [comment]
The tasks can report anything that can be sent to an Appendable.  The PrintWriter Appendable accepts characters, arrays of characters, and things that can be converted to arrays of characters.  Your task can write JSON 'String's to it if you wish.  The user may not thank you, if he runs it on the console.

There is no GUI code yet, so it can do what it will with the output.  Curator receives and Appendable, which can do anything with the stuff appended to it.  That's why I chose such an abstract interface.

The CLI task runner already has an output option -- see the -r option.  At the moment, output is discarded if there is no -r; '-r -' results in console output; '-r /path/to/a/file' results in file output.  There is no such thing as a "CLI task" -- the task only knows that it is to do something to some DSpaceObject and perhaps report its activity, but not what happens to the report.

Queued tasks:  the queues are drained by the CLI tool, so you have the same options as you have with immediate runs.  If someday we have a GUI interface for "run queued tasks" then it can supply any sort of Appendable to handle the output.

If you want to see a sample of how output might be preserved when there is no CLI running, take a look at WorkflowCuratorService, which runs deep inside the submission workflow.  The patch collects task output as files in '[dspace]/logs/curation-YYYYMMDDThhmmssSSS.log'.

**Comment by mwoodiupui** [comment]
I would suggest that turning reports into JSON, or HTML, or ASN.1, or whatever, is the responsibility of the code which invokes the Curator.  It may be that we need to invent a kind of language for expressing structure, and then the method signatures can be amended to accept Appendable<UnitOfOurLanguage>.  We could then provide some rendering methods to turn a UnitOfOurLanguage into flat text, markup, serialized structures in a database....

**Comment by mwoodiupui** [comment]
Again I would consider log file naming, or whether there is a file at all, to be the responsibility of the code that instantiates the Curator.  I'm treating reporting as a property of the Curator, not of the tasks that it happens to be running.  Have the caller pass an appropriate report sink, whatever that entails.

As I read the original intent, a task was meant to be unaware of its environment beyond the boundaries of DSpace.  It is given a DSO; it can manipulate that DSO or in fact any part of the DSpace data model; it is given a way to report on its work; it may be given some parameters.  That's all.  It is indirectly coupled to the environment by the invoking code, through a Curator instance which hides *from the task* the details of how that task is managed.

**Comment by mwoodiupui** [comment]
Take another look at what Curator does.  It runs a list of tasks over a tree of DSOs.  The output of all of those task invocations is *one report*.  Thus tasks should *not* be able to influence the output destination or they will step on each other.

**Comment by mwoodiupui** [comment]
If there is a need for a task to know what kind of output (plain text, JSON, etc.) is wanted, see #2181 and also "task properties".

File: dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java [hunk 1]
```diff
@@ -118,14 +124,23 @@ public boolean doCuration(Context c, BasicWorkflowItem wfi)
...
                 basicWorkflowItemService.update(c, wfi);
                 return false;
             } else {
-                return curate(curator, c, wfi);
+                Date now = GregorianCalendar.getInstance().getTime();
+                SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddThhmmssSSS");
+                String filename = String.format("curation-%s.log", sdf.format(now));
+                Path logPath = Paths.get(
+                        configurationService.getProperty("dspace.dir"), "logs", filename);
+                try (PrintWriter reporter = new PrintWriter(logPath.toFile())) {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: Why are we "manually" managing this log file? Is there a way to use log4j for this?  It seems a bit odd to me to do this outside of our `log4j.properties` file, and hardcode it to a specific name/date format.  Ideally, I'd want to manage this alongside all other DSpace logs.

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: This code yields a separate log file for each run; log4j would collect them all in daily aggregate files.  I'd be happy to discuss how we want to capture the output of curation runs that happen without an admin present.

I should bear in mind that WorkflowCuratorService would always be applied to single Items.  OTOH there is no limit to how much a task may report for a single object.

Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: As a sidenote... I *believe* some amount of curation task information already is "discovered" by log4j and ends up in the `dspace.log` file.  I know I've seen curation tasks trigger output in the `dspace.log` before.  So, if we are now thinking a separate curation.log file is useful (and in my opinion it seems useful), we may want to redirect that output as well.

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: I do wonder whether we want to distinguish between logging stuff that the code mutters about its own operation, and tasks' reports on the objects on which they operate.  Something about the word "report" makes me think it should be separate.  Maybe I should have directed the files to [DSpace]/reports.  (Come to think of it, if this code continues to make its own files, the destination should be configurable.)

The main dspace.log is such a rag-bag of unconnected stuff that I often ponder how to break it apart into multiple streams with more focus.

Anyway it should not be difficult to whip up an Appendable implementation that just stuffs the characters into a log category.  Perhaps a specific, documented category that can be directed to a separate log appender.  "workflow.${collection UUID}.${taskID}" or some such.

I wonder if it is worth bothering to worry about what happens when two workflows are progressing in parallel -- what would that look like, coming out a single reporting channel?  Worst case would be two (or more!) GUI batch loads with tasks attached to Step 1, so we have multiple curations per second going off immediately on submission.  This is why the separate report filenames are dated down to the millisecond....

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: Aside:  I've come to think that the CLI and the webapp.s should each have its own log configuration and log files.

Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: If you want reports to the millisecond, and want separate tasks in separate logs, I'd assume you can configure that in log4j.  In log4j you can configure the log file naming structure...you can also configure which individual Java packages/classes log to which log files.

I'm just worried here about generating unexpected logs outside of log4j. Plus, the lack of configurability here could become a pain point to some, based on how they manage their logs (as some folks do use log analyzers to "watch" for errors/warnings and unexpected logs using a hardcoded naming convention/path can mean information is overlooked or missed)

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: @terrywbrady seems to have some specific ideas about how task reports would be used.  Maybe he can help us sort this out.

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: I would say that errors and warnings should go to dspace.log, just as with the rest of the code.  I wouldn't use the report stream for that.  I'd use reporting to say "I successfully operated on object X and here's what I have to say as a result."

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: We could punt:  define interface o.d.curation.ReportWriter, and 2-3 stock implementation plugins.  A LogReportWriter, a FileReportWriter, maybe a CommandReportWriter....

Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: @mwoodiupui : It's worth mentioning that (as of DSpace 6.x) the location of DSpace log files is actually also configurable within log4j (and *is not controlled by the dspace.cfg in any way*):
https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace/config/log4j.properties#L19
(This change came in as part of the new Configuration System, and was somewhat necessitated by it -- as log4j could not "sync up" with Apache Commons Configuration)

This `log.dir` in log4j is autopopulated during `ant` deployment, but could obviously be modified after deployment as needed.

So, if an institution changes their log4j configuration to store the DSpace logs under `/var/log/dspace/` (or wherever), this _should_ work fine....except that, hardcoded log locations obviously won't be aware of that, and will continue to log elsewhere.

This is why I'm pushing a bit hard on using log4j.  If we don't use log4j, then logs could end up in unexpected locations, since all logs are now configured via log4j.

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: Ah, that's a plain bug.  I didn't remember `log.dir` or I should have used it.  But I think it was a mistake to use the 'log' directory at all for files I don't think of as "logs".

Anyway I think we need a few more discussants to work out how reports are going to be used and how they ought to be organized, and then see how to put them in the right place by the right path.

Comment (terrywbrady) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: @mwoodiupui , here are my thoughts about task reports.

I would like to create Curation Tasks that enforce expectations for the contents of a collection such as the relationships between bitstreams and metadata.

For instance, I would like to scan a community and verify that every item has a type and that type matches a value from a list of standard types.  Based on the type of an item, I would then like to ensure that the bitstreams conform to my expectations.  For instance, if I have a type of "Thesis", I would expect to see a PDF, a thumbnail, and some extracted text.  

For items with a type of "Audio" or "Video", I would have a different set of standards to apply.

For any items that do not conform to these rules, I would like to report (1)a URL to the item in question and (2)a description of the non-conforming attributes of the item.

Because a curation task can be run in the background, it would be possible to set up a robust set of checks that would be run against each collection a few times per year.

Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java]: Over the weekend, I sketched up untested LogReporter and FileReporter plugins together with the plugin plumbing.  They aren't hard to write, and now there is a Reporter interface (needed by the PluginService) for a guide.  I haven't pushed the changes up yet, awaiting discussion to see what is really needed.

The FileReporter writes per-run files to [DSpace]/reports (configurable).  The LogReporter writes whole lines (strings ending in newline) to log records through the category "curation.$taskID" so they can be picked off to a separate curation log or even separate logs per taskID.

**Comment by mwoodiupui** [comment]
I think the blank log records are fixed -- I couldn't reproduce them after adding a test to skip flushing the record buffer if it's empty.

File: dspace-api/src/main/java/org/dspace/curate/CurationCli.java [hunk 2]
```diff
@@ -147,9 +152,17 @@ public static void main(String[] args) throws Exception {
         }
 
         Curator curator = new Curator();
-        if (reporterName != null) {
-            curator.setReporter(reporterName);
+        OutputStream reporter;
+        if (null == reporterName) {
```
Comment (terrywbrady) [inline dspace-api/src/main/java/org/dspace/curate/CurationCli.java]: Update the CLI usage information to describe the values that may be passed in for reporter.  

Clarify that a relative path name or absolute path name can be specified.  Clarify the intention of the "-" value.

In my testing, output seemed to be written to the log as well as to the file.  Should curation output be suppressed from the log when a ReportWriter is active?

**Review by terrywbrady — CHANGES_REQUESTED** [review]
Requesting a minor change to the comments in the .cfg file and in the CLI usage text.

These changes will be very handy.  I wish these were present in 5x and 6x.

File: dspace-api/src/main/java/org/dspace/curate/CurationCli.java (hunk 2 above)
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/curate/CurationCli.java]: If a task calls e.g. Logger.info() then that goes straight to the log.  If a task calls Curator.report() then that should go straight to the Reporter that was set and not to the log.  A task may have different things to say in the log and in a report.  Of course a task may send the same text out both channels, separately.

**Review by terrywbrady — APPROVED** [review]

**Comment by mwoodiupui** [comment]
Rebased to cure conflicts.

File: dspace/config/modules/submission-curation.cfg [hunk 3]
```diff
@@ -5,4 +5,8 @@
 # to the scheduling of curation tasks during submission.        #
 #---------------------------------------------------------------#
 # Scan for viruses
-submission-curation.virus-scan = false
\ No newline at end of file
+submission-curation.virus-scan = false
+
+# Report serializer plugin.  Uncomment exactly one, or configure your own.
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace/config/modules/submission-curation.cfg]: Done.

**Review by benbosman — APPROVED** [review]
This is looking good to me. The only thing I'm wondering is whether we can somehow further verify the various Reporters in the Integration Test.
It can be useful to also verify in the IT whether there's any misuse of System out using something similar to 
https://stackoverflow.com/a/1119559/4078814. And alternatively, the logger can be set to System out using that procedure and it can be verified whether the data is actually logged to System out.

A similar process can perhaps also be used to verify the FileReporter and LogReporter.

**Comment by mwoodiupui** [comment]
Rebased to fix a conflict.
